### PR TITLE
Group the optimizer components into a package

### PR DESCRIPTION
**Pull request by ivy-zhou**
Summary:
@tianyu-l asked for a dedicated folder for the optimizer components, and both
@fegin and @tianyu-l pointed out that `checkpoint_utils.py` is not an optimizer
module even though three of its four public functions are optimizer-only.

Split it on that line and group the result:

  components/optimizer.py                 -> components/optimizer/optimizer.py
  components/lr_scheduler.py              -> components/optimizer/lr_scheduler.py
  components/checkpoint_utils.py:optimizer state helpers
                                          -> components/optimizer/utils.py

`init_optim_state`, `get_flat_optim_state_dict`, and `load_flat_optim_state_dict`
all take a `torch.optim.Optimizer`, so they move. `canonical_fqn(name: str, ...)`
does not: it strips activation-checkpoint wrapper segments from a dotted FQN and
is a checkpointing concern. It stays in `components/checkpoint_utils.py` here and
moves to `components/checkpointer/utils.py` in the next change, which is the one
that creates that folder, so it only ever moves once.

Nothing outside the package has to change its imports:

- `components/optimizer/__init__.py` re-exports `OptimizersContainer`,
  `ParamGroupConfig`, `default_adamw`, `register_moe_load_balancing_hook`, and
  `LRSchedulersContainer`, so `torchtitan.components.optimizer` keeps working and
  the scheduler now comes from the same short path as the container.
- `components/lr_scheduler.py` and `components/checkpoint_utils.py` stay as
  forwarding modules for their old names.

EMA has no module today, so there is nothing to move for it yet.

Test Plan:
`pytest tests/unit_tests/test_checkpoint.py tests/unit_tests/test_state_dict_keys.py
tests/unit_tests/test_lr_scheduler.py tests/unit_tests/test_optimizer_param_groups.py`
passes (68 tests) at this commit, in a venv matching the CPU CI job.

`torchtitan.trainer` and `torchtitan.components.optimizer` both import standalone
at this commit. `flake8` and `ufmt` (pinned black 22.12.0 / usort 1.0.5, matching
`.pre-commit-config.yaml`) are clean on the changed files.

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/pytorch/torchtitan/pull/4140).
* #4172
* #4058
* #4141
* __->__ #4140

### CODE REVIEW

**Comment by ivy-zhou** [comment]
cc: @tianyu-l I'm also happy to move this into an optimizer specific dir if you want? I just wanted to make this because it was getting a bit confusing what should actually go in the components/checkpointer folder.

File: torchtitan/components/optimizer/utils.py
Comment (tianyu-l) [inline torchtitan/components/optimizer/utils.py:1]: only concern is https://github.com/pytorch/torchtitan/pull/4141#discussion_r3779288149

**Review by tianyu-l — COMMENTED** [review]
> I'm also happy to move this into an optimizer specific dir if you want?

Yeah I think we can move lr scheduler, optimizer, and maybe ema into a dedicated folder. Not urgent though.

**Review by fegin — COMMENTED** [review]
`canonical_fqn` is also used by rl trainer.py. And it is not related to optimizer. Just saying that this function doesn't fit into `optimizer_utils` naming

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/components/checkpoint_utils.py [hunk 1]
```diff
@@ -50,190 +29,8 @@ def canonical_fqn(name: str, prefixes: tuple[str, ...] = _WRAPPER_PREFIXES) -> s
...
-        parts = key[len(prefix_dot) :].split(".")
-        for part in parts[:-1]:
-            current = current.setdefault(part, {})
-        current[parts[-1]] = value
-    return result
+from .optimizer.utils import (  # noqa: E402
+    get_flat_optim_state_dict,
+    init_optim_state,
+    load_flat_optim_state_dict,
+)
```
Comment (fegin) [inline torchtitan/components/checkpoint_utils.py:36]: Do we still need to re-import? Or should we just let the downstream usage to directly import `optimizer.utils`?

File: torchtitan/components/lr_scheduler.py [hunk 2]
```diff
@@ -4,244 +4,10 @@
 # This source code is licensed under the BSD-style license found in the
```
Comment (fegin) [inline torchtitan/components/lr_scheduler.py:4]: Why do we still keep this file? Who is using this? If it is the torchtitan-package-internal usage, we should just change them to use the new path.

**Review by fegin — COMMENTED** [review]
The BC-compatible shim doesn't seem worth the effort. Just change the downstream usages.

File: torchtitan/components/lr_scheduler.py (hunk 2 above)
Comment (ivy-zhou) [reply torchtitan/components/lr_scheduler.py:4]: Absolutely yeah, discussed offline that bc it's a large number of changes to the files, we can address in a separate [PR](https://github.com/pytorch/torchtitan/pull/4172)
